### PR TITLE
histogram: Handle changes of the ZeroThreshold and the Schema

### PR DESCRIPTION
**Pull request by beorn7**
__This is just for the sparsehistogram branch. Nothing to see here.__

### CODE REVIEW

File: model/histogram/float_histogram.go [hunk 1]
```diff
@@ -647,80 +884,111 @@ func (b FloatBucket) String() string {
...
+		if i.schema == i.targetSchema {
+			// Don't need to test the next bucket for mergeability
+			// if we have no schema change anyway.
+			break mergeLoop
 		}
-		span = r.spans[r.spansIdx]
-		r.currIdx += span.Offset
 	}
+	// Skip buckets before absoluteStartValue.
+	// TODO(beorn7): Maybe do something more efficient than this recursive call.
```
Comment (codesome) [inline model/histogram/float_histogram.go:959]: During the first pass of Next(), we can zoom through the initial buckets whose target idxs' bounds are below absoluteStartValue.

**Review by codesome — APPROVED** [review]
LGTM. I did not review the test cases; I trust that they are right.

File: model/histogram/float_histogram.go [hunk 2]
```diff
@@ -647,80 +884,111 @@ func (b FloatBucket) String() string {
...
-		Index:          r.currIdx,
+// targetIdx returns the bucket index within i.targetSchema for the given bucket
+// index within i.schema.
+func (i *floatBucketIterator) targetIdx(idx int32) int32 {
+	if i.schema == i.targetSchema {
+		// Fast path for the common case. The below would yield the same
+		// result, just with more effort.
+		return idx
 	}
+	return ((idx - 1) >> (i.schema - i.targetSchema)) + 1
```
[the remark this answers is not in the document]
Comment (beorn7) [reply model/histogram/float_histogram.go:991]: Yeah, quite some brain cycles went into it. The only thing that gives me confidence are the tests. And drawings on my whiteboard… (o:

File: model/histogram/float_histogram.go (hunk 1 above)
Comment (beorn7) [reply model/histogram/float_histogram.go:959]: That's what I tried initially. But there were weird corner cases that I found hard to resolve. The recursive appeared to me the logically cleanest solution. Perhaps let's try to tweak this in a separate PR (that's why I left the TODO).

**Pull request merged**